File: neetcode all and misc questions/Graphs/island_perimeter.py
"""
https://leetcode.com/problems/island-perimeter/
"""
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        """
        1) If the cell is 1, count the neighbors. For each neighboring cell, if the neighboring cell is 0, or if 
            it's out of bounds, increment the perimeter
            
            0 1 0 0 
            1 1 1 0
            0 1 0 0
            1 1 0 0

            [0, 1] has a water cell to the left and right, and is out of bounds on the top, so increment by 3
            [1, 0] has a water cell above, below, and is out of bounds on the left side, so increment by 3
            [1, 1] is connected by islands cells all around it, so don't increment by anything
            [1, 2] is connected by water cells above, below and to the right, so increment by 3
            [2, 1] is connected by water cells to the left and to the right, increment by 2
            [3, 0] is connected by water cells above, and out bounds to the left and below, increment by 3
            [3, 1] is connected by water cells to the right, and out of bounds on the bottom, increment by 2
            3 + 3 + 3 + 2 + 3 + 2 = 16
                    
            O(N^2) Time 
        """
        def inBounds(i, j, grid):
            return 0 <= i < len(grid) and 0 <= j < len(grid[0])
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        perimeter = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    for d in directions:
                        x, y = d
                        newX, newY = i + x, j + y
                        if inBounds(newX, newY, grid) and grid[newX][newY] == 0 or (not inBounds(newX, newY, grid)):
                            perimeter += 1
        return perimeter
                    

File: node_js_ds_and_algos_implementations/graphs/island_perimeter.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    islandPerimeter(grid) {
        /*
        because there is only one island, we can just assume
        that every grid[i][j] === 1 is connected to each other,
        so we don't need DFS

        in terms of the perimeter, for each cell, we need to know whether
        its on the boundaries of the grid, and/or if there's a cell === 0
        that's 4 directional to it. In that case, we would add 1 for each side
        that's either on a boundary of the grid or a water cell.
    
        O(N*M) time
        O(1) space
        */
        let total = 0
        let directions = [[0,1],[0,-1],[1,0],[-1,0]]
        function inBounds(i,j){
            return 0 <= i && i < grid.length && 0 <= j && j < grid[0].length
        }
        for (let i = 0; i < grid.length; ++i){
            for (let j = 0; j < grid[0].length; ++j){
                if (grid[i][j] === 1){
                    for (let [x,y] of directions){
                        let newX = x+i
                        let newY = y+j
                        // if the neighboring cell is out of bounds,
                        // this boundary is included in the perimeter
                        if (!inBounds(newX,newY)){
                            ++total
                        }
                        // if neighboring cell is a water cell,
                        // boundary is included in the perimeter
                        else if (inBounds(newX, newY) && grid[newX][newY] === 0){
                            ++total
                        }
                    }
                }
            }
        }
        return total
    }
}
